src/simplification/Simplifier.ts

```diff
@@ -73,57 +48,80 @@ export class Simplifier {
-        models = [...models, ...simplifiedProperties.newModels];
+      if (simplifiedProperties !== undefined) {
+        model.properties = simplifiedProperties;
       }
       
       const simplifiedAdditionalProperties = simplifyAdditionalProperties(schema, this, model);
-      if (simplifiedAdditionalProperties.newModels !== undefined) {
-        models = [...models, ...simplifiedAdditionalProperties.newModels];
-      }
-      if (simplifiedAdditionalProperties.additionalProperties !== undefined) {
-        model.additionalProperties = simplifiedAdditionalProperties.additionalProperties;
+      if (simplifiedAdditionalProperties !== undefined) {
+        model.additionalProperties = simplifiedAdditionalProperties;
       }
 
-      if (this.options.allowInheritance) {
-        const simplifiedExtends = simplifyExtend(schema, this);
-        if (simplifiedExtends.newModels !== undefined) {
-          models = [...models, ...simplifiedExtends.newModels];
-        }
-        if (simplifiedExtends.extendingSchemas !== undefined) {
-          model.extend = simplifiedExtends.extendingSchemas;
-        }
+      const simplifiedExtends = simplifyExtend(schema, this);
+      if (simplifiedExtends !== undefined) {
+        model.extend = simplifiedExtends;
       }
 
       const enums = simplifyEnums(schema);
       if (enums !== undefined && enums.length > 0) {
-        if (model.enum) {
-          model.enum = [...model.enum, ...enums];
-        } else {
-          model.enum = enums;
-        }
+        model.enum = enums;
       }
 
       const required = simplifyRequired(schema);
       if (required !== undefined) {
         model.required = required;
       }
     }
+    this.ensureModelsAreSplit(model);
+    const modelsToReturn = Object.values(this.iteratedModels);
+    //Add models which have not been iterated before
+    if (isModelObject(model) && this.iteratedModels[`${model.$id}`] === undefined) {
+      this.iteratedModels[`${model.$id}`] = model;
+    }
+    return [model, ...modelsToReturn];
+  }
 
-    //Always ensure the model representing the input schema to be in index 0. 
-    models = [model, ...models];
-    return models;
+  /**
+   * Split up all models which should and use ref instead.
+   * 
+   * @param model to ensure are split
+   * @param models which are already split
+   */
+  private ensureModelsAreSplit(model: CommonModel) {
+    /**
+    * This function splits up a model if needed and add the new model to the list of models.
+    * 
+    * @param model check if it should be split up
+    * @param models which have already been split up
+    */
+    const splitModels = (model: CommonModel) : CommonModel => {
+      if (isModelObject(model)) {
+        const switchRootModel = new CommonModel();
+        switchRootModel.$ref = model.$id;
+        this.iteratedModels[`${model.$id}`] = model;
+        return switchRootModel;
+      }
+      return model;
+    };
```

Comment:
This function should be moved to outside of parent function but in `Simplifier` scope.

```ts
private ensureModelsAreSplit(model: CommonModel) { ... }

/**
    * This function splits up a model if needed and add the new model to the list of models.
    * 
    * @param model check if it should be split up
    * @param models which have already been split up
    */
    const splitModels = (model: CommonModel): CommonModel => {
      if (isModelObject(model)) {
        const switchRootModel = new CommonModel();
        switchRootModel.$ref = model.$id;
        this.iteratedModels[`${model.$id}`] = model;
        return switchRootModel;
      }
      return model;
    };
```